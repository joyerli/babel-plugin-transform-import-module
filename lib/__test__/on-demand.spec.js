const babel = require('@babel/core');
const _ = require('lodash');
const nu = require('@upman/node-utils');
const plugin = require('../index');


function testUnit(source, options) {
  const { code } = babel.transform(source, {
    filename: 'path/to/file.vue',
    sourceType: 'module',
    plugins: [[plugin, {
      library: {
        'element-ui': (componentName, alias) => [{
          moduleId: alias || componentName,
          source: `element-ui/lib/${_.kebabCase(componentName)}`,
        }, {
          source: `path/to/${_.kebabCase(componentName)}.css`,
        }],
      },
      ...options,
    }]],
  });
  expect(code).toMatchSnapshot();
}

function resetMessage(message) {
  return _.last(message.split(':'));
}

test('导入单个el组件', () => {
  testUnit(`
    import { Button } from 'element-ui';
    console.log(Button);
  `);
});
test('别名', () => {
  testUnit(`
  import { Alert as alert } from 'element-ui'
  console.log(alert);
  `);
});
test('导入多个组件', () => {
  testUnit(`
    import { Alert, Button } from 'element-ui'
    console.log(Alert, Button);
  `);
});
test('别名跟普通导入', () => {
  testUnit(`
    import { Alert as alert, Button } from 'element-ui'
    console.log(alert, Button);
  `);
});
test('多行导入', () => {
  testUnit(`
    import { Alert } from 'element-ui'
    import { Button } from 'element-ui'
    console.log(alert, Button);
  `);
});
test('导入文件', () => {
  testUnit(`
    import 'element-ui';
    console.log('=====');
  `, {
    preventFullImport: false,
  });
});
test('命令空间导入', () => {
  testUnit(`
    import * as Element from 'element-ui';
    console.log(Element);
  `, {
    preventFullImport: false,
  });
});
test('默认导入', () => {
  testUnit(`
    import * as Element from 'element-ui';
    console.log(Element);
  `, {
    preventFullImport: false,
  });
});
test('禁止全局导入之导入文件', () => {
  const { message } = nu.catchErrorSync(() => {
    testUnit(`
      import Element from 'element-ui';
    `);
  });
  expect(resetMessage(message)).toMatchSnapshot();
});
test('禁止全局导入之命令空间导入', () => {
  const { message } = nu.catchErrorSync(() => {
    testUnit(`
      import Element from 'element-ui';
    `);
  });
  expect(resetMessage(message)).toMatchSnapshot();
});
test('禁止全局导入之默认导入', () => {
  const { message } = nu.catchErrorSync(() => {
    testUnit(`
      import * as Element from 'element-ui';
      console.log(Element);
    `);
  });
  expect(resetMessage(message)).toMatchSnapshot();
});
test('允许导出', () => {
  testUnit(`
    console.log(Element);
    export { default } from 'element-ui';
  `, {
    preventExport: false,
  });
});
test('禁止导出', () => {
  const { message } = nu.catchErrorSync(() => {
    testUnit(`
      console.log(Element);
      export { default } from 'element-ui';
    `);
  });
  expect(resetMessage(message)).toMatchSnapshot();
});
