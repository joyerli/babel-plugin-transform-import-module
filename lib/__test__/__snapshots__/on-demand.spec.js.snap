// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`允许导出 1`] = `
"console.log(Element);
export { default } from 'element-ui';"
`;

exports[`别名 1`] = `
"import alert from \\"element-ui/lib/alert\\";
import \\"path/to/alert.css\\";
console.log(alert);"
`;

exports[`别名跟普通导入 1`] = `
"import alert from \\"element-ui/lib/alert\\";
import \\"path/to/alert.css\\";
import Button from \\"element-ui/lib/button\\";
import \\"path/to/button.css\\";
console.log(alert, Button);"
`;

exports[`命令空间导入 1`] = `
"import * as Element from 'element-ui';
console.log(Element);"
`;

exports[`多行导入 1`] = `
"import Alert from \\"element-ui/lib/alert\\";
import \\"path/to/alert.css\\";
import Button from \\"element-ui/lib/button\\";
import \\"path/to/button.css\\";
console.log(alert, Button);"
`;

exports[`导入单个el组件 1`] = `
"import Button from \\"element-ui/lib/button\\";
import \\"path/to/button.css\\";
console.log(Button);"
`;

exports[`导入多个组件 1`] = `
"import Alert from \\"element-ui/lib/alert\\";
import \\"path/to/alert.css\\";
import Button from \\"element-ui/lib/button\\";
import \\"path/to/button.css\\";
console.log(Alert, Button);"
`;

exports[`导入文件 1`] = `
"import 'element-ui';
console.log('=====');"
`;

exports[`禁止全局导入之命令空间导入 1`] = `
" Not allowed import all module
[0m [90m 1 | [39m[0m
[0m[31m[1m>[22m[39m[90m 2 | [39m      [36mimport[39m [33mElement[39m from [32m'element-ui'[39m[33m;[39m[0m
[0m [90m   | [39m      [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m
[0m [90m 3 | [39m    [0m"
`;

exports[`禁止全局导入之导入文件 1`] = `
" Not allowed import all module
[0m [90m 1 | [39m[0m
[0m[31m[1m>[22m[39m[90m 2 | [39m      [36mimport[39m [33mElement[39m from [32m'element-ui'[39m[33m;[39m[0m
[0m [90m   | [39m      [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m
[0m [90m 3 | [39m    [0m"
`;

exports[`禁止全局导入之默认导入 1`] = `
" Not allowed import all module
[0m [90m 1 | [39m[0m
[0m[31m[1m>[22m[39m[90m 2 | [39m      [36mimport[39m [33m*[39m as [33mElement[39m from [32m'element-ui'[39m[33m;[39m[0m
[0m [90m   | [39m      [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m
[0m [90m 3 | [39m      console[33m.[39mlog([33mElement[39m)[33m;[39m[0m
[0m [90m 4 | [39m    [0m"
`;

exports[`禁止导出 1`] = `
" Not allowed due to preventExport setting
[0m [90m 1 | [39m[0m
[0m [90m 2 | [39m      console[33m.[39mlog([33mElement[39m)[33m;[39m[0m
[0m[31m[1m>[22m[39m[90m 3 | [39m      [36mexport[39m { [36mdefault[39m } from [32m'element-ui'[39m[33m;[39m[0m
[0m [90m   | [39m      [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m
[0m [90m 4 | [39m    [0m"
`;

exports[`默认导入 1`] = `
"import * as Element from 'element-ui';
console.log(Element);"
`;
